src/OrleansProviders/Streams/Common/PooledCache/IObjectPool.cs

```diff
@@ -35,17 +35,10 @@ public abstract class PooledResource<T> : IDisposable
         private IObjectPool<T> pool;
 
         /// <summary>
-        /// Pooled resource that is from the provided pool
+        /// The pool to return this resource to upon disposal.
+        /// A pool must set this property upon resource allocation.
         /// </summary>
-        /// <param name="pool"></param>
-        protected PooledResource(IObjectPool<T> pool)
-        {
-            if (pool == null)
-            {
-                throw new ArgumentNullException("pool");
-            }
-            this.pool = pool;
-        }
```

Comment:
Looking at the [Dispose method in this class](https://github.com/dotnet/orleans/pull/1937/files#diff-86104cd38875be8f0defd13603d19a8bR61) I believe casting `this as T` would only work if you if you make the constrait `T : PooledResource<T>` (instead of `T : class`).
That way you enforce it in the compiler, and you should also be able to use explicit cast instead of safe cast (which would not be really safe, but error prone instead).